starter/class/connect-four.js

```diff
@@ -21,24 +21,88 @@ class ConnectFour {
     Screen.setGridlines(true);
 
     // Replace this with real commands
-    Screen.addCommand('t', 'test command (remove)', ConnectFour.testCommand);
+    // Screen.addCommand('t', 'test command (remove)', ConnectFour.testCommand);
+    Screen.addCommand('up', 'moves cursor up', this.cursor.up);
+    Screen.addCommand('Down', 'moves cursor down', this.cursor.down);
+    Screen.addCommand('left', 'move cursor left', this.cursor.left);
+    Screen.addCommand('right', 'moves cursor right', this.cursor.right);
+    Screen.addCommand('return', "places a move in the current position", this.placeMove)
 
     this.cursor.setBackgroundColor();
     Screen.render();
   }
+  // Return 'X' if player X wins
+  // Return 'O' if player O wins
+  // Return 'T' if the game is a tie
+  // Return false if the game has not ended
 
-  // Remove this
-  static testCommand() {
-    console.log("TEST COMMAND");
+  placeMove = () => {
+    if (Screen.grid[this.cursor.row][this.cursor.col] === " ") {
+      Screen.setTextColor(this.cursor.row, this.cursor.col, 'white');
+      Screen.setGrid(this.row, this.col, this.playerTurn);
```

Comment:
2. Now we're here, as this is the only place in which it is called.

We can verify this is the right direction we're going if you read your stack trace very carefully:

```
1.  Workspace/connect-four/starter/class/screen.js:95
2.     Screen.grid[row][col] = char;
3.                           ^
4. 
5.  TypeError: Cannot set property 'undefined' of undefined
6.     at Function.setGrid (Workspace/connect-four/starter/class/screen.js:95:27)
7.     at Command.placeMove [as action] (Workspace/connect-four/starter/class/connect-four.js:42:14)
8.     at Command.execute (Workspace/connect-four/starter/class/command.js:10:10)
9.     at ReadStream.<anonymous> (Workspace/connect-four/starter/class/screen.js:80:35)
10.    at ReadStream.emit (events.js:400:28)
11.    at emitKey (Workspace/connect-four/starter/node_modules/keypress/index.js:406:12)
12.    at ReadStream.onData (Workspace/connect-four/starter/node_modules/keypress/index.js:48:14)
13.    at ReadStream.emit (events.js:400:28)
14.    at addChunk (internal/streams/readable.js:293:12)
15.    at readableAddChunk (internal/streams/readable.js:267:9)
```

Let's break it down line by line:
1. This is where we're having the error, at the exact line.
2. The error line that is the issue
3. The cursor, pointing to the exact variable in which the issue is starting at (in this case, it's pointing to `row`)
4. <blank>
5. It indicates that "row" is `undefined. So at this point the alarms should go off in your head to say "how did it become undefined?"
6. Now it beings the"stack tracing", meaning it's explaining how its tracing the code back through this callable stack, starting with the first execution call [at screen.js:95:27](https://github.com/HandoCodeCoaching/connect-four/blob/a86343033577948578546d2574dcc9fd2a75dae7/starter/class/screen.js#L95)
7. The trace now takes you to [`placeMove` on line 42](https://github.com/HandoCodeCoaching/connect-four/blob/a86343033577948578546d2574dcc9fd2a75dae7/starter/class/connect-four.js#L42):

```js
  placeMove = () => {
    if (Screen.grid[this.cursor.row][this.cursor.col] === " ") {
      Screen.setTextColor(this.cursor.row, this.cursor.col, 'white');
      Screen.setGrid(this.row, this.col, this.playerTurn);
```


I think this should get you to figure out the solution.

Reply:
Thank you for this write-up, very helpful to confirm that I understand the errors that I'm looking at.